airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -49,33 +49,49 @@ def _get_dep_statuses(self, ti, session, dep_context):
             yield self._passing_status(reason="The task had a dummy trigger rule set.")
             return
 
-        # TODO(unknown): this query becomes quite expensive with dags that have many
-        # tasks. It should be refactored to let the task report to the dag run and get the
-        # aggregates from there.
-        qry = (
-            session
-            .query(
-                func.coalesce(func.sum(
-                    case([(TI.state == State.SUCCESS, 1)], else_=0)), 0),
-                func.coalesce(func.sum(
-                    case([(TI.state == State.SKIPPED, 1)], else_=0)), 0),
-                func.coalesce(func.sum(
-                    case([(TI.state == State.FAILED, 1)], else_=0)), 0),
-                func.coalesce(func.sum(
-                    case([(TI.state == State.UPSTREAM_FAILED, 1)], else_=0)), 0),
-                func.count(TI.task_id),
+        successes, skipped, failed, upstream_failed, done = 0, 0, 0, 0, 0
+        if dep_context.finished_tasks is None:
+            qry = (
+                session
+                .query(
+                    func.coalesce(func.sum(
+                        case([(TI.state == State.SUCCESS, 1)], else_=0)), 0),
+                    func.coalesce(func.sum(
+                        case([(TI.state == State.SKIPPED, 1)], else_=0)), 0),
+                    func.coalesce(func.sum(
+                        case([(TI.state == State.FAILED, 1)], else_=0)), 0),
+                    func.coalesce(func.sum(
+                        case([(TI.state == State.UPSTREAM_FAILED, 1)], else_=0)), 0),
+                    func.count(TI.task_id),
```

Comment:
May be another user for State.finished() (and makes me think that UPSTREAM_FAILED should be in there)

```python
                 .query(*[
                   func.coalesce(func.sum(case([(TI.state == state, 1)], else_=0)), 0)
                   for state in State.finished()
                 ])
                 .filter(
                    TI.dag_id == ti.dag_id,
                    TI.task_id.in_(ti.task.upstream_task_ids),
                    TI.execution_date == ti.execution_date,
                    TI.state.in_(State.finished()),
                 )
```

Reply:
Yes, almost the same here, I will add UPSTREAM_FAILED and we will see what happen